File: programming-in-the-past/fortran/caesar.f90
! Entrypoint for the program
program caesar
    ! We need to make sure that all variable types are explicitly defined
    implicit none

    ! We will need 3 strings, the original, the output for encrypt, and the output for decrypt
    character(31) :: original
    character(31) :: encryptOutput
    character(31) :: decryptOutput

    original = 'This is a test string from Alan'

    ! Call encrypt on the original string with a shift of 8
    call encrypt(original, 31, -27, encryptOutput)
    print *, encryptOutput

    ! Decrypt the encrypted value with the shift of 8
    call decrypt(encryptOutput, 31, -27, decryptOutput)
    print *, decryptOutput

    ! Call the solve subroutine
    call solve('HAL', 3, 26)
    
contains

    ! Subroutine to encrypt a string of size stringLength by shiftAmount and store the result in outputString
    subroutine encrypt(inputString, stringLength, shiftAmount, outputString)
        implicit none

        ! stringLength is needed so the type of inputString and outputString can be explicitly defined by their length
        ! shiftAmount is the amount to shift the string by
        integer, intent(in) :: stringLength, shiftAmount

        ! inputString will not be modified and is a character type of size stringLength
        character(stringLength), intent(in) :: inputString
        ! outputString will be modified and has the same length as inputString
        character(stringLength) :: outputString

        ! newShift will be used to make sure the shift amount is between -25 and 25
        ! index will be used to iterate through the string character by character
        ! charValue is the ASCII value of a given character
        ! diff will be used to determine if the wraparound calculation is needed
        integer :: newShift, index, charValue, diff

        ! Start off by setting outputString to the same value is inputString
        outputString = inputString

        ! Convert the shift amount to be between -25 and 25
        newShift = mod(shiftAmount, 26)

        ! Iterate through the string one character at a time
        ! Looping and modifying individual characters found at the link below
        ! http://computer-programming-forum.com/49-fortran/4075a24f74fcc9ce.htm
        do index = 1, len(outputString)
            ! Get the original character value at the given position
            charValue = ichar(outputString(index:index))

            ! Check to see if it is a lowercase letter ('a' - 'z')
            if (charValue >= 97 .and. charValue <= 122) then
                ! If so, make it capital by subtracting 32 ('a' - 'A')
                charValue = charValue - 32
            endif

            ! Make sure the character is a letter ('A' - 'Z') because already converted to uppercase
            if (charValue >= 65 .and. charValue <= 90) then
                ! Perform the shift by the given amount
                charValue = charValue + newShift

                ! diff is how much over the charater is relative to 'Z' (90)
                diff = charValue - 90

                ! If it went over, wraparound is needed
                if (diff > 0) then
                    ! Finish the wraparound by adding the diff to 'A' (65)
                    ! The -1 is needed because a diff of 1 means that the character is 1 beyond 'Z', which will be 'A'
                    charValue = 65 + diff - 1
                else
                    ! diff is how much under the charater is relative to 'A' (65)
                    diff = 65 - charValue

                    ! If it went under, wraparound is needed
                    if (diff > 0) then
                        ! Finish the wraparound by subtracting the diff from 'Z' (90)
                        ! The +1 is needed because a diff of 1 means that the character is 1 beyond 'A', which will be 'Z'
                        charValue = 90 - diff + 1
                    endif
                endif

                ! Update the output string at the index with the new character value
                ! If the character was not originally a letter, then no change is needed, which is why this is inside of the if block
                outputString(index:index) = char(charValue)
            endif
        enddo
    end subroutine encrypt

    ! Subroutine to decrypt a string of size stringLength that has been encrypted by shiftAmount and store the result in outputString
    subroutine decrypt(inputString, stringLength, shiftAmount, outputString)
        implicit none

        ! stringLength is needed so the type of inputString and outputString can be explicitly defined by their length
        ! shiftAmount is the amount to unshift by
        integer, intent(in) :: stringLength, shiftAmount

        ! inputString will not be modified and is a character type of size stringLength
        character(stringLength), intent(in) :: inputString
        ! outputString will be modified and has the same length as inputString
        character(stringLength) :: outputString

        ! newShift will be used to make sure the shift amount is between -25 and 25
        ! index will be used to iterate through the string character by character
        ! charValue is the ASCII value of a given character
        ! diff will be used to determine if the wraparound calculation is needed
        integer :: newShift, index, charValue, diff

        ! Start off by setting outputString to the same value is inputString
        outputString = inputString

        ! Convert the shift amount to be between -25 and 25
        newShift = mod(shiftAmount, 26)

        ! Iterate through the string one character at a time
        ! Looping and modifying individual characters found at the link below
        ! http://computer-programming-forum.com/49-fortran/4075a24f74fcc9ce.htm
        do index = 1, len(outputString)
            ! Get the original character value at the given position
            charValue = ichar(outputString(index:index))

            ! Check to see if it is a lowercase letter ('a' - 'z')
            if (charValue >= 97 .and. charValue <= 122) then
                ! If so, make it capital by subtracting 32 ('a' - 'A')
                charValue = charValue - 32
            endif

            ! Make sure the character is a letter ('A' - 'Z') because already converted to uppercase
            if (charValue >= 65 .and. charValue <= 90) then
                ! Perform the unshift by the given amount
                charValue = charValue - newShift

                ! diff is how much under the charater is relative to 'A' (65)
                diff = 65 - charValue

                ! If it went under, wraparound is needed
                if (diff > 0) then
                    ! Finish the wraparound by subtracting the diff from 'Z' (90)
                    ! The +1 is needed because a diff of 1 means that the character is 1 beyond 'A', which will be 'Z'
                    charValue = 90 - diff + 1
                else
                    ! diff is how much over the charater is relative to 'Z' (90)
                    diff = charValue - 90

                    ! If it went over, wraparound is needed
                    if (diff > 0) then
                        ! Finish the wraparound by adding the diff to 'A' (65)
                        ! The -1 is needed because a diff of 1 means that the character is 1 beyond 'Z', which will be 'A'
                        charValue = 65 + diff - 1
                    endif
                endif

                ! Update the output string at the index with the new character value
                ! If the character was not originally a letter, then no change is needed, which is why this is inside of the if block
                outputString(index:index) = char(charValue)
            endif
        enddo
    end subroutine decrypt

    ! Subroutine to solve a caesar cipher of the input string of size stringLength with a set number of tries (maxShiftValue)
    subroutine solve(inputString, stringLength, maxShiftValue)
        implicit none

        ! stringLength is needed so the type of inputString and outputString can be explicitly defined by their length
        ! shiftAmount is the amount to unshift by
        integer, intent(in) :: stringLength, maxShiftValue

        ! inputString will not be modified and is a character type of size stringLength
        character(stringLength), intent(in) :: inputString

        ! outputString is a temporary variable that will be used as the result from each solve attempt
        character(stringLength) :: outputString

        ! shiftAmount is the current amount being shifted
        integer :: shiftAmount

        ! Try all shift amounts from the max down to 0
        do shiftAmount = maxShiftValue, 0, -1
            ! Call the decrypt subroutine with the current shiftAmount and store in outputString
            call decrypt(inputString, stringLength, -shiftAmount, outputString)

            ! Print out the results
            ! Formatting strings found at the link below
            ! https://pages.mtu.edu/~shene/COURSES/cs201/NOTES/chap05/format.html
            print '(a,I0,a,a)', 'Caesar ', shiftAmount, ': ', outputString
        enddo
    end subroutine solve
end program caesar

File: programming-in-the-past/pascal/caesar.pas
program Caesar;
(* Function that encrypts the input string by the shift amount *)
function Encrypt(inStr : string; shiftAmount : integer) : string;
var
    curChar: Char; (* The current character being worked with *)
    charValue: integer; (* The ASCII value of the current character *)
    diff: integer; (* The difference between the current character and A and Z to detect wraparound *)
    idx: integer; (* The position in the string we are at *)
begin
    (* Start off by normalizing the shift amount *)
    shiftAmount := shiftAmount mod 26;
    Encrypt := inStr;

    (* Pascal uses 1-based indexing *)
    idx := 1;
    (* Loop through each character in the string *)
    for curChar in Encrypt do
    begin
        (* Get the initial character ASCII code *)
        charValue := ord(curChar);

        (* Convert the character to uppercase if it is not already *)
        if ((charValue >= 97) and (charValue <= 122)) then
            charValue := charValue - 32;

        (* Caesar cipher only impacts letters, so check that conditon first *)
        if ((charValue >= 65) and (charValue <= 90)) then
        begin
            (* Perform the shift *)
            charValue := charValue + shiftAmount;

            (* Check wraparound on the Z end *)
            diff := charValue - 90;
            if diff > 0 then
                (* Perform the wraparound (-1 is needed because 91 should be 65 (A)) *)
                charValue := 65 + diff - 1
            else
            begin
                (* Check the low end of the range *)
                diff := 65 - charValue;
                if diff > 0 then
                    (* Perform the wraparound (+1 is needed because 64 should be 90 (Z)) *)
                    charValue := 90 - diff + 1;
            end;
            (* Update the string accordingly *)
            Encrypt[idx] := chr(charValue);
        end;
        (* Increment the index for our own reference *)
        idx := idx + 1;
    end;
end;

(* Function that decrypts the input string by the shift amount *)
function Decrypt(inStr : string; shiftAmount : integer) : string;
var
    curChar: Char; (* The current character being worked with *)
    charValue: integer; (* The ASCII value of the current character *)
    diff: integer; (* The difference between the current character and A and Z to detect wraparound *)
    idx: integer; (* The position in the string we are at *)
begin
    (* Start off by normalizing the shift amount *)
    shiftAmount := shiftAmount mod 26;
    Decrypt := inStr;

    (* Pascal uses 1-based indexing *)
    idx := 1;
    (* Loop through each character in the string *)
    for curChar in Decrypt do
    begin
        (* Get the initial character ASCII code *)
        charValue := ord(curChar);

        (* Convert the character to uppercase if it is not already *)
        if ((charValue >= 97) and (charValue <= 122)) then
            charValue := charValue - 32;

        (* Caesar cipher only impacts letters, so check that conditon first *)
        if ((charValue >= 65) and (charValue <= 90)) then
        begin
            (* Perform the shift *)
            charValue := charValue - shiftAmount;

            (* Check wraparound on the Z end *)
            diff := charValue - 90;
            if diff > 0 then
                (* Perform the wraparound (-1 is needed because 91 should be 65 (A)) *)
                charValue := 65 + diff - 1
            else
            begin
                (* Check the low end of the range *)
                diff := 65 - charValue;
                if diff > 0 then
                    (* Perform the wraparound (+1 is needed because 64 should be 90 (Z)) *)
                    charValue := 90 - diff + 1;
            end;
            (* Update the string accordingly *)
            Decrypt[idx] := chr(charValue);
        end;
        (* Increment the index for our own reference *)
        idx := idx + 1;
    end;
end;

(* Procedure to solve a Caesar cipher *)
procedure solve(inputString : string; maxShiftAmount : integer);
var
    curShift: integer; (* The current shift amount  *)
begin
    (* Normalize the max shift amount being used *)
    maxShiftAmount := abs(maxShiftAmount);
    if maxShiftAmount > 26 then
        maxShiftAmount := maxShiftAmount mod 26;

    (* Loop through everything *)
    for curShift := maxShiftAmount downto 0 do
        (* Perform the decrypt and print the result *)
        writeln('Caesar ', curShift, ': ', Decrypt(inputString, -curShift));
end;

var
    originalString: string; (* The original string *)
    encryptOut: string; (* The output of encrypt *)
    decryptOut: string; (* The output of decrypt *)
begin
    originalString := 'This is a test string from Alan';
    writeln('Alan tests:');
    encryptOut := Encrypt(originalString, 8);
    decryptOut := Decrypt(encryptOut, 8);
    writeln(encryptOut);
    writeln(decryptOut);
    solve('HAL', 26);
    
    writeln();
    writeln('Encrypt and decrypt tests:');
    (* Test negative shift *)
    encryptOut := Encrypt(originalString, -1);
    decryptOut := Decrypt(encryptOut, -1);
    writeln(encryptOut);
    writeln(decryptOut);
    (* Test modulus *)
    encryptOut := Encrypt(originalString, 27);
    decryptOut := Decrypt(encryptOut, 27);
    writeln(encryptOut);
    writeln(decryptOut);
    (* Test empty string *)
    encryptOut := Encrypt('', 7);
    decryptOut := Decrypt(encryptOut, 7);
    writeln(encryptOut);
    writeln(decryptOut);
    (* Test no letters *)
    encryptOut := Encrypt('1234567890!@#$%^&*(){}', 7);
    decryptOut := Decrypt(encryptOut, 7);
    writeln(encryptOut);
    writeln(decryptOut);

    writeln();
    writeln('Solve tests:');
    (* Test absolute value *)
    solve('HAL', -26);
    writeln();
    (* Test modulus *)
    solve('HAL', 30);
end.